text/2018-11-12-unified-key-format.md

```diff
@@ -0,0 +1,93 @@
+  turns out that these implementations won't work in real world scenario. For example, it is
+  not sufficient to decode the Golang format by only dealing with hex escapes `\x..` [\[1\]][1]
+  and decode the Protobuffer format by only dealing with oct escapes `\...` [\[2\]][2].
+
+- Hard to recognize different parts manually: For hex format, it is hard to extract different parts
+  from it manually.
+
+- Hard to be used in shells: For Protobuffer format and Go format, because there are escape
+  characters U+005C (\\) in the output, we need to do an escape (convert `\` to `\\`) when passing
+  it as a command line parameter in shells. As a result, current documentations are suggesting
+  wrong usages [\[3\]][3] that *happens* to work under specially
+  crafted code, or won't work [\[4\]][4] actually.
+
+- Contains ambiguous multi-bytes: For Go format, it directly outputs Unicode character if there is
+  a valid unicode sequence. For example, hex `00E6B58B01FF` will be printed as `\x00测\x01\xff` in
+  Go format. This may result in ambiguous bytes represents in different environments when such
+  characters are printed.
+
+To overcome above issues, we need a new human readable format and use it among all TiDB software
+stack components, which provides these feature:
+
+- Easy to recognize different parts after encoding by human.
+
+- Easy to implement encoding and decoding. Better to have a lot of existing libraries.
+
+- No escape character U+005C (\\) after encoding so that it can be easily used in shells.
+
+- No multi-byte characters after encoding to avoid encoding issues.
+
+# Detailed Design
+
+This RFC proposes to use hex format like `74000000000000001C5F7200000000000000FA` as unified a human
+readable format for keys:
+
+- Each byte must be encoded to a 2 digit hex in upper case.
+
+- Encoded hex string must be concatenated together without any other characters.
+
+- Decoder must support hex string in either lower-case or upper-case.
+
+Although hex format has some defects as discussed in the motivation section above, it is still
+preferred over other formats (see alternatives section) with great advantages in simplicity and
+uniformity. In addition, it is very easy to be implemented in all languages and does not introduce
+escape character U+005C (\\).
+
+In hex format, we are not easy to recognize special parts like `r`, `_t` and `_i`. This can be
+solved by other utility functions that further operates on this key format, or just memorizing
+the hex representation of these magic bytes.
+
+# Drawbacks
+
+- We need to memorize hex representation of magic bytes, or implement additional helper tools.
+
+- It is not compatible with current formats and we need to write format converters.
+
+# Alternatives
+
+Other binary-to-text encodings like Base64, Percent-Encoding and punycode are investigated. For
+Base64 and punycode, it is harder to read magic bytes manually. For Percent-Encoding, its
```

Comment:
This is also a drawback of our implementation, so this shouldn't disqualify them. We need a better reason. :)